Unable to open already-opened shared memory on macOS Aarch64

Sample log:
```
com.skaggsm.sharedmemory.SharedMemoryTest > Given a shared memory reference, when another reference to the same name is reopened, then both are usable FAILED
    java.lang.AssertionError: Property failed after 1 attempts

    Repeat this test by using seed -1077834448824682046

    Caused by RuntimeException: Unusual failure of shm_open with O_RDWR! Message: Permission denied
        at com.skaggsm.sharedmemory.SharedMemoryTest$1$4.invokeSuspend(SharedMemoryTest.kt:69)
        at com.skaggsm.sharedmemory.SharedMemoryTest$1$4.invoke(SharedMemoryTest.kt)
        at com.skaggsm.sharedmemory.SharedMemoryTest$1$4.invoke(SharedMemoryTest.kt)

        Caused by:
        java.lang.AssertionError: Property failed after 1 attempts

        Repeat this test by using seed -1077834448824682046

        Caused by RuntimeException: Unusual failure of shm_open with O_RDWR! Message: Permission denied
            at com.skaggsm.sharedmemory.SharedMemoryTest$1$4.invokeSuspend(SharedMemoryTest.kt:69)
            ... 2 more

            Caused by:
            java.lang.RuntimeException: Unusual failure of shm_open with O_RDWR! Message: Permission denied
                at com.skaggsm.sharedmemory.posix.SharedMemoryPOSIX.<init>(SharedMemoryPOSIX.java:40)
                at com.skaggsm.sharedmemory.SharedMemory.getSharedMemory(SharedMemory.java:34)
                at com.skaggsm.sharedmemory.SharedMemoryTest$1$4$1.invokeSuspend(SharedMemoryTest.kt:75)
                at com.skaggsm.sharedmemory.SharedMemoryTest$1$4$1.invoke(SharedMemoryTest.kt)
                at com.skaggsm.sharedmemory.SharedMemoryTest$1$4$1.invoke(SharedMemoryTest.kt)
                at io.kotest.property.internal.ProptestKt$proptest$3$2.invokeSuspend(proptest.kt:32)
                at io.kotest.property.internal.ProptestKt$proptest$3$2.invoke(proptest.kt)
                at io.kotest.property.internal.ProptestKt$proptest$3$2.invoke(proptest.kt)
                at io.kotest.property.internal.TestKt.test(test.kt:21)
                at io.kotest.property.internal.ProptestKt.proptest(proptest.kt:31)
                at io.kotest.property.PropertyTest1Kt.checkAll(propertyTest1.kt:16)
                ... 3 more
```

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -28,8 +28,11 @@ repositories {
 }
 
 dependencies {
-    implementation("net.java.dev.jna:jna:4.4.0")
-    implementation("net.java.dev.jna:platform:3.4.0")
+    compileOnly("net.java.dev.jna:jna:4.4.0")
+    compileOnly("net.java.dev.jna:platform:3.4.0")
+
+    runtimeOnly("net.java.dev.jna:jna:5.8.0")
+    runtimeOnly("net.java.dev.jna:jna-platform:5.8.0")
 
     testImplementation("net.java.dev.jna:jna:5.8.0")
     testImplementation("net.java.dev.jna:jna-platform:5.8.0")

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -31,6 +31,9 @@ dependencies {
     implementation("net.java.dev.jna:jna:4.4.0")
     implementation("net.java.dev.jna:platform:3.4.0")
 
+    testImplementation("net.java.dev.jna:jna:5.8.0")
+    testImplementation("net.java.dev.jna:jna-platform:5.8.0")
+
     testImplementation(kotlin("stdlib-jdk8"))
     testImplementation("org.junit.jupiter:junit-jupiter:5.7.2")
 

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -6,7 +6,7 @@ import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 plugins {
     java
     `maven-publish`
-    kotlin("jvm") version "1.5.10"
+    kotlin("jvm") version "1.5.21"
     id("org.shipkit.shipkit-auto-version") version "1.+"
     id("org.shipkit.shipkit-changelog") version "1.+"
     id("org.shipkit.shipkit-github-release") version "1.+"
@@ -34,11 +34,11 @@ dependencies {
     testImplementation(kotlin("stdlib-jdk8"))
     testImplementation("org.junit.jupiter:junit-jupiter:5.7.2")
 
-    testImplementation("io.kotest:kotest-runner-junit5:4.6.0")
-    testImplementation("io.kotest:kotest-property:4.6.0")
-    testImplementation("io.kotest:kotest-assertions-core:4.6.0")
+    testImplementation("io.kotest:kotest-runner-junit5:4.6.1")
+    testImplementation("io.kotest:kotest-property:4.6.1")
+    testImplementation("io.kotest:kotest-assertions-core:4.6.1")
 
-    testImplementation("org.slf4j:slf4j-nop:2.0.0-alpha1")
+    testImplementation("org.slf4j:slf4j-nop:2.0.0-alpha2")
 }
 
 java {

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.1.1-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/src/main/java/com/skaggsm/sharedmemory/posix/LibC.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/LibC.java
@@ -2,13 +2,14 @@ package com.skaggsm.sharedmemory.posix;
 
 import com.sun.jna.Library;
 import com.sun.jna.Native;
+import com.sun.jna.Platform;
 
 /**
  * Created by Mitchell Skaggs on 5/15/2019.
  */
 
 public interface LibC extends Library {
-    LibC INSTANCE = Native.loadLibrary("c", LibC.class);
+    LibC INSTANCE = Native.loadLibrary(Platform.C_LIBRARY_NAME, LibC.class);
 
     int getuid();
 

--- a/src/main/java/com/skaggsm/sharedmemory/posix/LibRT.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/LibRT.java
@@ -8,14 +8,14 @@ import com.sun.jna.Pointer;
 /**
  * Created by Mitchell Skaggs on 5/15/2019.
  */
-interface LibRT extends Library {
-    LibRT INSTANCE = Native.loadLibrary(getRTLibraryName(), LibRT.class);
+public interface LibRT extends Library {
+    LibRT INSTANCE = getInstance();
 
-    static String getRTLibraryName() {
+    static LibRT getInstance() {
         if (Platform.isMac())
-            return Platform.C_LIBRARY_NAME;
+            return Native.loadLibrary(Platform.C_LIBRARY_NAME, MacOSLibRT.class);
         else
-            return "rt";
+            return Native.loadLibrary("rt", LibRT.class);
     }
 
     int shm_open(String name, int oFlag, int mode);
@@ -25,4 +25,12 @@ interface LibRT extends Library {
     Pointer mmap(Pointer addr, long length, int prot, int flags, int fd, long offset);
 
     int munmap(Pointer addr, long length);
+
+    interface MacOSLibRT extends LibRT {
+        default int shm_open(String name, int oFlag, int mode) {
+            return shm_open(name, oFlag, new Object[]{mode});
+        }
+
+        int shm_open(String name, int oFlag, Object... mode);
+    }
 }

--- a/src/main/java/com/skaggsm/sharedmemory/posix/LibRT.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/LibRT.java
@@ -13,7 +13,7 @@ interface LibRT extends Library {
 
     static String getRTLibraryName() {
         if (Platform.isMac())
-            return null;
+            return Platform.C_LIBRARY_NAME;
         else
             return "rt";
     }

--- a/src/main/java/com/skaggsm/sharedmemory/posix/MMAN.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/MMAN.java
@@ -11,7 +11,7 @@ abstract class MMAN {
     static final int PROT_READ;
     static final int PROT_WRITE;
     static final int MAP_SHARED;
-    static final Pointer MAP_FAILED = new Pointer(-1);
+    static final Pointer MAP_FAILED = Pointer.createConstant(-1);
 
     static {
         switch (Platform.getOSType()) {

--- a/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
@@ -40,7 +40,7 @@ public class SharedMemoryPOSIX implements SharedMemory {
                 RuntimeException t = new RuntimeException("Unusual failure of shm_open with O_RDWR! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
                 int shmUnlinkCode = LibRT.INSTANCE.shm_unlink(name);
                 if (shmUnlinkCode < 0 && Native.getLastError() != 2 /*ENOENT*/)
-                    throw new RuntimeException("Failed shm_unlink! Message: " + LibC.INSTANCE.strerror(Native.getLastError()), t);
+                    throw new RuntimeException("Unusual secondary failure of shm_unlink after failed shm_open! Message: " + LibC.INSTANCE.strerror(Native.getLastError()), t);
                 else
                     throw t;
             }

--- a/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
@@ -36,9 +36,17 @@ public class SharedMemoryPOSIX implements SharedMemory {
 
         fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR, S_IRUSR | S_IWUSR);
         if (fileDescriptor < 0) {
+            if (Native.getLastError() != 2 /*ENOENT*/) {
+                RuntimeException t = new RuntimeException("Unusual failure of shm_open with O_RDWR! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
+                int shmUnlinkCode = LibRT.INSTANCE.shm_unlink(name);
+                if (shmUnlinkCode < 0 && Native.getLastError() != 2 /*ENOENT*/)
+                    throw new RuntimeException("Failed shm_unlink! Message: " + LibC.INSTANCE.strerror(Native.getLastError()), t);
+                else
+                    throw t;
+            }
             fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
             if (fileDescriptor < 0)
-                throw new RuntimeException("Failed shm_open with O_CREAT! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
+                throw new RuntimeException("Failed shm_open with O_RDWR | O_CREAT! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
             hasOwnership = true;
         } else
             hasOwnership = false;

--- a/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
@@ -17,13 +17,13 @@ import static com.sun.jna.Pointer.NULL;
 
 public class SharedMemoryPOSIX implements SharedMemory {
 
-    private int fileDescriptor;
-    private Pointer memory;
-    private long size;
-    private String name;
+    private final int fileDescriptor;
+    private final Pointer memory;
+    private final long size;
+    private final String name;
+    private final boolean hasOwnership;
 
     private boolean closed;
-    private boolean hasOwnership;
 
     public SharedMemoryPOSIX(String name, long size) {
         this.size = size;
@@ -34,7 +34,7 @@ public class SharedMemoryPOSIX implements SharedMemory {
         if (this.name.length() > 31)
             System.err.printf("Name %s is probably too long for macOS (max 31 chars)%n", this.name);
 
-        fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR, S_IRUSR | S_IWUSR);
+        int fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR, S_IRUSR | S_IWUSR);
         if (fileDescriptor < 0) {
             if (Native.getLastError() != 2 /*ENOENT*/) {
                 RuntimeException t = new RuntimeException("Unusual failure of shm_open with O_RDWR! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
@@ -50,6 +50,7 @@ public class SharedMemoryPOSIX implements SharedMemory {
             hasOwnership = true;
         } else
             hasOwnership = false;
+        this.fileDescriptor = fileDescriptor;
 
         if (hasOwnership) {
             int ftruncateCode = LibC.INSTANCE.ftruncate(fileDescriptor, size);
@@ -85,9 +86,6 @@ public class SharedMemoryPOSIX implements SharedMemory {
             if (shmUnlinkCode < 0)
                 throw new RuntimeException("Failed shm_unlink! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
         }
-
-        memory = null;
-        fileDescriptor = -1;
 
         closed = true;
     }

--- a/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
@@ -17,7 +17,6 @@ import static com.sun.jna.Pointer.NULL;
 
 public class SharedMemoryPOSIX implements SharedMemory {
 
-    private final int fileDescriptor;
     private final Pointer memory;
     private final long size;
     private final String name;
@@ -47,20 +46,21 @@ public class SharedMemoryPOSIX implements SharedMemory {
             fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
             if (fileDescriptor < 0)
                 throw new RuntimeException("Failed shm_open with O_RDWR | O_CREAT! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
-            hasOwnership = true;
-        } else
-            hasOwnership = false;
-        this.fileDescriptor = fileDescriptor;
 
-        if (hasOwnership) {
             int ftruncateCode = LibC.INSTANCE.ftruncate(fileDescriptor, size);
             if (ftruncateCode < 0)
                 throw new RuntimeException("Failed ftruncate! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
-        }
+            hasOwnership = true;
+        } else
+            hasOwnership = false;
 
         memory = LibRT.INSTANCE.mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fileDescriptor, 0);
         if (memory.equals(MAP_FAILED))
             throw new RuntimeException("Failed mmap! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
+
+        int closeCode = LibC.INSTANCE.close(fileDescriptor);
+        if (closeCode < 0)
+            throw new RuntimeException("Failed to close file descriptor! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
     }
 
     @Override
@@ -76,10 +76,6 @@ public class SharedMemoryPOSIX implements SharedMemory {
         int munmapCode = LibRT.INSTANCE.munmap(memory, size);
         if (munmapCode < 0)
             throw new RuntimeException("Failed munmap! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
-
-        int closeCode = LibC.INSTANCE.close(fileDescriptor);
-        if (closeCode < 0)
-            throw new RuntimeException("Failed close! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
 
         if (hasOwnership) {
             int shmUnlinkCode = LibRT.INSTANCE.shm_unlink(name);

--- a/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
+++ b/src/main/java/com/skaggsm/sharedmemory/posix/SharedMemoryPOSIX.java
@@ -37,20 +37,21 @@ public class SharedMemoryPOSIX implements SharedMemory {
         fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR, S_IRUSR | S_IWUSR);
         if (fileDescriptor < 0) {
             fileDescriptor = LibRT.INSTANCE.shm_open(this.name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+            if (fileDescriptor < 0)
+                throw new RuntimeException("Failed shm_open with O_CREAT! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
             hasOwnership = true;
-        }
-        if (fileDescriptor < 0)
-            throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+        } else
+            hasOwnership = false;
 
         if (hasOwnership) {
             int ftruncateCode = LibC.INSTANCE.ftruncate(fileDescriptor, size);
             if (ftruncateCode < 0)
-                throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+                throw new RuntimeException("Failed ftruncate! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
         }
 
         memory = LibRT.INSTANCE.mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fileDescriptor, 0);
         if (memory.equals(MAP_FAILED))
-            throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+            throw new RuntimeException("Failed mmap! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
     }
 
     @Override
@@ -65,16 +66,16 @@ public class SharedMemoryPOSIX implements SharedMemory {
 
         int munmapCode = LibRT.INSTANCE.munmap(memory, size);
         if (munmapCode < 0)
-            throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+            throw new RuntimeException("Failed munmap! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
 
         int closeCode = LibC.INSTANCE.close(fileDescriptor);
         if (closeCode < 0)
-            throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+            throw new RuntimeException("Failed close! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
 
         if (hasOwnership) {
             int shmUnlinkCode = LibRT.INSTANCE.shm_unlink(name);
             if (shmUnlinkCode < 0)
-                throw new RuntimeException(LibC.INSTANCE.strerror(Native.getLastError()));
+                throw new RuntimeException("Failed shm_unlink! Message: " + LibC.INSTANCE.strerror(Native.getLastError()));
         }
 
         memory = null;

